src/daemon/plugin.c

```diff
@@ -2094,9 +2094,13 @@ static void set_default_resource_attributes(metric_family_t *fam) {
 }
 
 EXPORT int plugin_dispatch_metric_family(metric_family_t const *fam) {
-  if ((fam == NULL) || (fam->metric.num == 0)) {
+  if (fam == NULL) {
+    ERROR("plugin_dispatch_metric_family: fam == NULL");
     return EINVAL;
   }
+  if (fam->metric.num == 0) {
+    return 0;
+  }
```

Comment:
While I'm fine with removing `if (fams[i].metric.num > 0)` checks from the _plugins_, I'm against collectd core just silently ignoring that.

In which situations such plugin behavior (trying to report 0 metrics) would not be a bug?

PS. checks for `metric.num` against zero are in:
```
$ grep "if.*metric.num .* 0" *.c
chrony.c:    if (fams[i].metric.num > 0) {
df.c:    if (families[i]->metric.num == 0) {
disk.c:    if (fams[i]->metric.num == 0) {
ipc.c:    if (fams[i].metric.num > 0) {
irq.c:  if (fam.metric.num > 0) {
olsrd.c:    if (metric_num > 0)
ping.c:    if (fams[i]->metric.num > 0) {
write_prometheus.c:  if (prom_fam->metric.num == 0)
write_prometheus.c:  if (fam->metric.num == 0) {
write_prometheus.c:    if (prom_fam->metric.num == 0) {
```

Reply:
The *disk plugin* is a good example: it defines a list of 8 metric families. Then architecture specific code adds metrics to some or all of these metric families. At the end, architecture agnostic code is calling `plugin_dispatch_metric_family` for all metric families. Previously it needed to check whether the family contained any metrics. Now this check is done centrally.

Overall I don't think that calling a function that accpets a list with an empty list is an error.

Reply:
The locations you called out are either simplified in thie PR (most), false positives (olsrd), or working as intended (write_prometheus).

Reply:
> Overall I don't think that calling a function that accepts a list with an empty list is an error.

At least it should not be silently ignored, it should be logged, as it could be a plugin error.

Duplicate metrics are logged (e.g. https://github.com/collectd/collectd/issues/4200), so IMHO missing metrics should also...

Reply:
I added a debug message. I really don't think we should log at any higher severity, this is working as intended.